File: exercism/ruby/queen-attack/queen_attack.rb
class Queens

  def initialize (white: [0, 3], black: [7, 3])
    raise ArgumentError if white == black

    @w_row, @w_file = white
    @b_row, @b_file = black
  end

  def white
    [@w_row, @w_file]
  end

  def black
    [@b_row, @b_file]
  end

  def to_s
    board = Array.new(8){ ["_"] * 8 }

    board[@w_row][@w_file] = "W"
    board[@b_row][@b_file] = "B"

    board.map{|row| row.join(" ") }.join("\n")
  end

  def attack?
    delta_row = (@w_row - @b_row).abs
    delta_file = (@w_file - @b_file).abs

    delta_row.zero? || delta_file.zero? || delta_row == delta_file
  end
end

File: exercism/elixir/queen-attack/queen_attack.exs
defmodule Queens do
  @typep position :: { integer, integer }
  @opaque t :: %__MODULE__{ black: position, white: position }
  defstruct black: nil, white: nil

  @doc """
  Creates a new set of Queens
  """
  @spec new(position, position) :: t
  def new(white \\ {0, 3}, black \\ {7, 3}) do
    if white == black, do: raise ArgumentError
    %__MODULE__{ white: white, black: black }
  end

  @doc """
  Gives a string reprentation of the board with
  white and black queen locations shown
  """
  @spec to_string(t) :: String.t
  def to_string(%__MODULE__{ white: white, black: black }) do
    Enum.map_join(0 .. 7, "\n", fn row ->
      Enum.map_join(0 .. 7, " ", fn col ->
        case { row, col } do
          ^white -> "W"
          ^black -> "B"
          _ -> "_"
        end
      end)
    end)
  end

  @doc """
  Checks if the queens can attack each other
  """
  @spec can_attack?(t) :: Boolean.t
  def can_attack?(%__MODULE__{ white: { wx, wy }, black: { bx, by } }) do
    dx = abs(wx - bx)
    dy = abs(wy - by)
    min(dx, dy) == 0 || dx == dy
  end
end
